File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/217-contains-duplicate-approach3.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func containsDuplicate(nums []int) bool {
	seen := make(map[int]struct{})
	for _, n := range nums {
		if _, ok := seen[n]; ok {
			return true
		}
		seen[n] = struct{}{}
	}
	return false
}

func runTests() {
	assert(containsDuplicate([]int{1, 2, 3, 1}) == true)
	assert(containsDuplicate([]int{1, 2, 3, 4}) == false)
	assert(containsDuplicate([]int{1, 1, 1, 3, 3, 4, 3, 2, 4, 2}) == true)
	assert(containsDuplicate([]int{}) == false)
	assert(containsDuplicate([]int{5}) == false)
	assert(containsDuplicate([]int{5, 5}) == true)
	fmt.Println("all tests pass")
}

func main() {
	runTests()
}

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/217-contains-duplicate-approach3.py
def contains_duplicate(nums: list[int]) -> bool:
    seen = set()
    for x in nums:
        if x in seen:
            return True
        seen.add(x)
    return False

assert contains_duplicate([1, 2, 3, 1]) == True
assert contains_duplicate([1, 2, 3, 4]) == False
assert contains_duplicate([1, 1, 1, 3, 3, 4, 3, 2, 4, 2]) == True
assert contains_duplicate([]) == False
assert contains_duplicate([5]) == False
assert contains_duplicate([5, 5]) == True
print("all tests pass")
